src/content/blog/maintaining-ddev-tests-contributor-training.md

```diff
@@ -0,0 +1,128 @@
+
+## What should a contributor work on?
+
+### Debugging broken tests
+
+Broken tests, of course, can be a result of broken code or broken tests. These often come up in the process of building a PR, and they need to be fixed, usually in the code, but sometimes it's the test that's at fault.
+
+In all cases, the first thing to do is to try to execute the test manually and locally, or perhaps just try to execute what the test was trying to do. For example, if the test fails doing an `app.Start()`, just try stepping through `app.Start()` in `ddev start`. 
+
+Demonstrations of stepping through a test in the debugger are below.
+
+### Flaky tests
+
+Flaky tests are the worst, because they may depend on the execution environment, or some other test may have polluted the environment, or they may happen in particular network situations.
+
+* Review the test output from Buildkite or GitHub workflow to try to understand what's actually going on. There may be an earlier error that's reported that causes the actual failure.
+* Try to run it manually, perhaps step-debugging and introducing slower execution.
+* Try to run it on Gitpod, which is a more similar environment.
+* Try to run it using [tmate](https://github.com/mxschmitt/action-tmate) on the GitHub workflow, which is as close as you can get to the real test environment. (Tmate is built into each of the DDEV test workflows.)
+
+### Fragile or brittle tests
+
+Fragile and brittle tests may be a result of specific expectations in the code, for example, looking for a particular string instead of a more general expectation. Sometimes these can be understood by looking through the test run. Often the test needs to be restructured, or the expectation can be turned into a regular expression instead of a specific literal expectation.
+
+### Slow tests
+
+The easiest place to look at slow tests is a full buildkite test run, which shows the timing for each test. In some cases, the timing is essential, for example, where we test every single database type and version and capability in [TestDdevAllDatabases](https://github.com/ddev/ddev/blob/507cfca2508b97786b80e8b8c83ea17f5c0fea20/pkg/ddevapp/ddevapp_test.go#L1652-L1893). But in many cases, we do too much and it doesn't all have to be done. We can *skip* the test in some environments (this one gets skipped on Lima and Colima). And we can do a limited selection of the most important test targets (this one does only `postgres:14`, `mariadb:10.11`, `mariadb:10.6`, `mariadb:10.4`, `mysql:8.0`, `mysql:5.7` when `GOTEST_SHORT=true`).
+
+* Does the test need to be run on this target system?
+* Do all the permutations need to be done?
+* Does DDEV need to be stopped and started over and over again?
+* Are we waiting too long for something, wasting time?
```

Comment:
Do these questions belong to the `Slow tests` paragraph? Because it seems strange to me that something is asked here without a headline.

Reply:
I guess these are the questions to be asked when working on slow tests.